library/src/main/kotlin/com/connectrpc/protocols/ConnectInterceptor.kt

```diff
@@ -95,13 +95,33 @@ internal class ConnectInterceptor(
                 val responseHeaders =
                     response.headers.filter { entry -> !entry.key.startsWith("trailer-") }
                 val compressionPool = clientConfig.compressionPool(responseHeaders[CONTENT_ENCODING]?.first())
+                val responseBody = try {
+                    compressionPool?.decompress(response.message.buffer) ?: response.message.buffer
+                } catch (e: Exception) {
+                    return@UnaryFunction HTTPResponse(
+                        code = Code.INTERNAL_ERROR,
+                        message = Buffer(),
+                        headers = responseHeaders,
+                        trailers = trailers,
+                        cause = ConnectException(
+                            code = Code.INTERNAL_ERROR,
+                            errorDetailParser = serializationStrategy.errorDetailParser(),
+                            message = e.message,
+                            exception = e,
+                        ),
+                        tracingInfo = response.tracingInfo,
+                    )
+                }
+                val message: Buffer
                 val (code, exception) = if (response.code != Code.OK) {
-                    val error = parseConnectUnaryException(code = response.code, response.headers, response.message.buffer)
+                    val error = parseConnectUnaryException(code = response.code, response.headers, responseBody)
+                    // We've already read the response body to parse an error - don't read again.
+                    message = Buffer()
```

Comment:
@erawhctim - This fixes the issue before where we'd both read a JSON error response and then read again for the message.